llvm/lib/CodeGen/MachinePipeliner.cpp

```diff
@@ -484,9 +499,35 @@ void MachinePipeliner::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.addRequired<MachineDominatorTree>();
   AU.addRequired<LiveIntervals>();
   AU.addRequired<MachineOptimizationRemarkEmitterPass>();
+  AU.addRequired<TargetPassConfig>();
   MachineFunctionPass::getAnalysisUsage(AU);
 }
 
+bool MachinePipeliner::runWindowScheduler(MachineLoop &L) {
+  MachineSchedContext Context;
+  Context.MF = MF;
+  Context.MLI = MLI;
+  Context.MDT = MDT;
+  Context.PassConfig = &getAnalysis<TargetPassConfig>();
```

Comment:
Where is this used?

Reply:
In the WindowScheduler, we use createMachineScheduler() through 'PassConfig' to call the target's custom MachineScheduler:
<img width="661" alt="image" src="https://github.com/llvm/llvm-project/assets/142874007/ed685e0b-2542-4001-81b4-5acbda9948cb">

Reply:
Additionally, two ScheduleDAGs are used here: one for analyzing the dependencies between all instructions after copying, and the other for scheduling instructions in the window. The difference is that the former does not need to consider register pressure.